File: backend/app/models/timeline.py
"""
Pydantic v2 models for the timeline API endpoints.

Provides response shapes for the three timeline endpoints:
  - GET /api/timeline            → TimelineResponse
  - GET /api/timeline/debates/{id} → DebateDetailResponse
  - GET /api/timeline/speakers/{debateId}/{speakerId} → SpeakerSummaryResponse
"""
from typing import Optional

from pydantic import BaseModel


class DebateSummary(BaseModel):
    """Lightweight debate entry nested inside a SessionCard."""

    id: str
    title: str
    speech_count: int


class SessionCard(BaseModel):
    """One parliamentary session shown in the timeline list."""

    id: str
    date: str                        # ISO date string "2026-04-07"
    chamber: str                     # "camera" or "senato"
    number: int
    recap: Optional[str] = None      # None if not yet generated
    debate_count: int
    vote_count: int
    speech_count: int
    debates: list[DebateSummary]


class TimelineResponse(BaseModel):
    """Paginated list of sessions for GET /api/timeline."""

    sessions: list[SessionCard]
    next_cursor: Optional[str] = None  # ISO date of the last session in this page
    has_more: bool


class PhaseInfo(BaseModel):
    """A single debate phase with speech count."""

    id: str
    title: str
    phase_type: Optional[str] = None
    speech_count: int


class VoteInfo(BaseModel):
    """One roll-call vote attached to a session."""

    id: str
    number: int
    subject: Optional[str] = None
    outcome: Optional[str] = None
    in_favor: Optional[int] = None
    against: Optional[int] = None
    abstained: Optional[int] = None


class ActInfo(BaseModel):
    """Parliamentary act discussed in a debate."""

    id: str
    title: Optional[str] = None
    type: Optional[str] = None


class SpeakerInfo(BaseModel):
    """One speaker in a debate, listed chronologically."""

    id: str
    first_name: str
    last_name: str
    party: Optional[str] = None
    speaking_role: Optional[str] = None
    is_government_member: bool = False
    speech_count: int
    phases: list[str]  # Phase titles where this speaker participated


class DebateDetailResponse(BaseModel):
    """Full debate detail for GET /api/timeline/debates/{id}."""

    id: str
    title: str
    recap: Optional[str] = None
    phases: list[PhaseInfo]
    speakers: list[SpeakerInfo]
    votes: list[VoteInfo]
    acts: list[ActInfo]


class SpeechText(BaseModel):
    """Full text of a single speech in a debate."""

    id: str
    text: str
    phase_title: Optional[str] = None


class SpeakerSummaryResponse(BaseModel):
    """Speaker speeches and optional AI summary for GET /api/timeline/speakers/{debateId}/{speakerId}."""

    summary: Optional[str] = None
    speech_count: int
    phases: list[str]
    speeches: list[SpeechText] = []

File: frontend/src/types/timeline.ts
/**
 * TypeScript interfaces for the Parliamentary Timeline feature.
 * These mirror the backend Pydantic models in backend/app/models/timeline.py
 */

export interface DebateSummary {
  id: string;
  title: string;
  speech_count: number;
}

export interface TimelineSession {
  id: string;
  date: string;           // ISO date "2026-04-07"
  chamber: string;        // "camera" | "senato"
  number: number;
  recap: string | null;
  debate_count: number;
  vote_count: number;
  speech_count: number;
  debates: DebateSummary[];
}

export interface TimelineResponse {
  sessions: TimelineSession[];
  next_cursor: string | null;
  has_more: boolean;
}

export interface PhaseInfo {
  id: string;
  title: string;
  phase_type: string | null;
  speech_count: number;
}

export interface VoteInfo {
  id: string;
  number: number;
  subject: string | null;
  outcome: string | null;
  in_favor: number | null;
  against: number | null;
  abstained: number | null;
}

export interface VoteParticipant {
  id: string;
  first_name: string;
  last_name: string;
  party: string | null;
  outcome: string; // "favor" | "against" | "absent"
}

export interface VotePartyBreakdown {
  party: string;
  favor: number;
  against: number;
  absent: number;
}

export interface VoteDetailResponse {
  id: string;
  number: number;
  subject: string | null;
  outcome: string | null;
  vote_type: string | null;
  in_favor: number | null;
  against: number | null;
  abstained: number | null;
  present: number | null;
  voters: number | null;
  majority: number | null;
  on_mission: number | null;
  breakdown: VotePartyBreakdown[];
  participants: VoteParticipant[];
}

export interface ActInfo {
  id: string;
  title: string | null;
  type: string | null;
}

export interface SpeakerInfo {
  id: string;
  first_name: string;
  last_name: string;
  party: string | null;
  speaking_role: string | null;
  is_government_member: boolean;
  speech_count: number;
  phases: string[];
}

export interface DebateDetailResponse {
  id: string;
  title: string;
  recap: string | null;
  phases: PhaseInfo[];
  speakers: SpeakerInfo[];
  votes: VoteInfo[];
  acts: ActInfo[];
}

export interface SpeechText {
  id: string;
  text: string;
  phase_title: string | null;
}

export interface SpeakerSummaryResponse {
  summary: string | null;
  speech_count: number;
  phases: string[];
  speeches: SpeechText[];
}

export interface TimelineFilters {
  chamber: string;    // "both" | "camera" | "senato"
  search: string;
  fromDate: string;   // ISO date or empty
  toDate: string;     // ISO date or empty
}
